### ISSUE TITLE
read stdin ? (like keyboard input)

### ISSUE THREAD

**Issue by mh-cc** [rra/remctl#33]
Hello,

I think this question is linked to https://github.com/rra/remctl/issues/10
And might relates to the first item in https://www.eyrie.org/~eagle/software/remctl/todo.html

So, if I understand correctly, could you confirm (or infirm) that in the current state:
- The remote command cannot read stdin, which means we cannot read keyboard input (for instance with `read` in bash or `input` in Python)
- The client cannot pipe data to the remote command's stdin. (for instance like described in https://github.com/rra/remctl/issues/10)

Thank you

**Comment by rra** [comment]
You can designate that one of the arguments to the remctl command is sent to standard input of the remote command (the `stdin` configuration parameter), and the client side can send that data incrementally rather than all at once. However, the very substantial caveats are that data sent this way is still buffered by the remctld server and only sent to the command once complete (so you can't do anything interactive), and the remctl command-line program doesn't know how to take one of the arguments from standard input (although if you're using bash as a shell you can use its functionality to do that and turn standard input into a command-line argument).

So, basically, interactive commands don't work through remctl. The protocol design doesn't let you send some data, get a response, send some more data, get another response, and so forth while still running the same command. Each send has to be a separate fork and exec on the server side. If you just need to route standard input to a command or send a lot of data that you don't want turned into a command-line argument, that works, but not actual interactivity.

`docs/protocol-v4` is a description of how that could be implemented, but none of the work has been done.

**Comment by mh-cc** [comment]
Hi @rra, thank you for the detailed answer. I'm not sure to fully understand what the solution to "route standard input to a command" could be. But I'll try to explain what I have in mind from my understanding and your answer.

On the server side, configure `stdin=2`
On the client side, do something like `remctl host command $(cat myfile)`
So the command can get and process the arguments data as `stdin`
That's right ?

However, I can't really predict how many arguments the command is going receive, and I'm not sure how much data I can provide to the command line in `myfile`.

**Comment by rra** [comment]
> On the server side, configure `stdin=2`

You probably want `stdin=last` so the last argument to `remctl` on the client (whatever number that is) is passed as standard input to the remote command.

> On the client side, do something like `remctl host command $(cat myfile)`

Right, except you'll want to quote that (`remctl host command "$(cat myfile)"`). I think there's some fancy bash construction that is better, but I don't remember what it is.

On Linux, you can pass about 50KB that way, if I remember correctly.

It obviously would be better for `remctl` to support reading standard input and turning it into an argument for the remote command, which is https://github.com/rra/remctl/issues/10 and which hasn't been implemented yet. Failing that, you can write a small Perl or Python (or Ruby or PHP) program using the library to do the equivalent, by reading the file in first and then passing the contents of the file as an argument to the remctl command.
